### ISSUE TITLE
JSON encoding issue when sending via out_http

### ISSUE THREAD

**Issue by jbhilton11** [fluent/fluent-bit#371]
When tailing /var/log/audit/audit.log using enriched (resolves user ids and other fields), auditd appends a codepoint ([U+001D](https://codepoints.net/U+001D), specifically) before appending the resolved fields. For instance, this could be what an auditd log looks like:

    type=PATH msg=audit(1364481363.243:24287): item=0 name="/etc/ssh/sshd_config" inode=409248 dev=fd:00 mode=0100600 ouid=0 ogid=0 rdev=00:00 obj=system_u:object_r:etc_t:s0^]UID=joe AUID=unset

In the audit log above, "^]" represents what is seen when viewing this in a visual editor (such as VI on Linux). When sending this output to an HTTP endpoint via out_http, I specify the host, port, URI, and format (json). When received on the HTTP endpoint, it cannot be parsed as JSON because the codepoint mentioned above has not been escaped correctly.

So, two-ish questions:
1. Can this be fixed, and if so when?
2. Until this is solved, does someone know of regex that can be used to ignore the codepoint altogether?

**Comment by edsiper** [comment]
@jbhilton11 would you please upload that file somewhere so I can reproduce locally ?

**Comment by jbhilton11** [comment]
@edsiper which file do you need? the audit.log file? or the configuration with parsers files? or all of those?

**Comment by edsiper** [comment]
the file with the unicode byte

**Comment by jbhilton11** [comment]
I'm actually not sure how I'd display that on here. I was able to reproduce the character that's added in by taking this specific log into vim and (when in insert mode) doing Ctrl+V and then typing x1d (the hex for the character) between "obj=system_u:object_r:etc_t:s0" and "UID=joe". That's what is added on when auditd configuration is set to "enriched". This causes the issue where it's not taken into consideration when converting to JSON before sending via out_http.

`type=PATH msg=audit(1364481363.243:24287): item=0 name="/etc/ssh/sshd_config" inode=409248 dev=fd:00 mode=0100600 ouid=0 ogid=0 rdev=00:00 obj=system_u:object_r:etc_t:s0 UID=joe AUID=unset`

**Comment by edsiper** [comment]
@jbhilton11 I did a test using the following auditd file:

https://bugzilla.redhat.com/attachment.cgi?id=1257213

the JSON generated is valid (and special characters is encoded properly), this is the JSON output I see the plugin is sending:

```json
[
  {
    "date": 1505439809.195425,
    "log": "type=USER_AVC msg=audit(1483362061.439:275): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='avc:  received policyload notice (seqno=2)  exe=\"/usr/lib/systemd/systemd\" sauid=0 hostname=? addr=? terminal=?'\u001d"
  },
  {
    "date": 1505439809.195437,
    "log": "type=AVC msg=audit(1483365315.456:288): avc:  denied  { write } for  pid=14838 comm=\"passwd\" path=\"/root/output-15305.txt\" dev=\"dm-1\" ino=2304 scontext=unconfined_u:unconfined_r:passwd_t:s0-s0:c0.c1023 tcontext=unconfined_u:object_r:admin_home_t:s0 tclass=file\u001d"
  },
  {
    "date": 1505439809.195438,
    "log": "type=AVC msg=audit(1483365315.456:288): avc:  denied  { write } for  pid=14838 comm=\"passwd\" path=\"/root/output-15305.txt\" dev=\"dm-1\" ino=2304 scontext=unconfined_u:unconfined_r:passwd_t:s0-s0:c0.c1023 tcontext=unconfined_u:object_r:admin_home_t:s0 tclass=file\u001d"
  },
  {
    "date": 1505439809.195438,
    "log": "type=SYSCALL msg=audit(1483365315.456:288): arch=c000003e syscall=59 success=yes exit=0 a0=1267c90 a1=1260070 a2=1270820 a3=7fff609a5cc0 items=0 ppid=1266 pid=14838 auid=0 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=pts0 ses=3 comm=\"passwd\" exe=\"/usr/bin/passwd\" subj=unconfined_u:unconfined_r:passwd_t:s0-s0:c0.c1023 key=(null)\u001dARCH=x86_64 SYSCALL=execve AUID=\"root\" UID=\"root\" GID=\"root\" EUID=\"root\" SUID=\"root\" FSUID=\"root\" EGID=\"root\" SGID=\"root\" FSGID=\"root\""
  },
  {
    "date": 1505439809.195439,
    "log": "type=USER_AVC msg=audit(1483365587.914:304): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='avc:  received policyload notice (seqno=3)  exe=\"/usr/lib/systemd/systemd\" sauid=0 hostname=? addr=? terminal=?'\u001dUID=\"root\" AUID=\"unset\" SAUID=\"root\""
  },
  {
    "date": 1505439809.195439,
    "log": "type=AVC msg=audit(1483365592.228:308): avc:  denied  { fowner } for  pid=29490 comm=\"haproxy\" capability=3  scontext=system_u:system_r:haproxy_t:s0 tcontext=system_u:system_r:haproxy_t:s0 tclass=capability\u001d"
  },
  {
    "date": 1505439809.19544,
    "log": "type=SYSCALL msg=audit(1483365592.228:308): arch=c000003e syscall=90 success=no exit=-1 a0=7ffdb5d8c610 a1=1ed a2=ffffffff a3=0 items=0 ppid=29486 pid=29490 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm=\"haproxy\" exe=\"/usr/sbin/haproxy\" subj=system_u:system_r:haproxy_t:s0 key=(null)\u001dARCH=x86_64 SYSCALL=chmod AUID=\"unset\" UID=\"root\" GID=\"root\" EUID=\"root\" SUID=\"root\" FSUID=\"root\" EGID=\"root\" SGID=\"root\" FSGID=\"root\""
  },
  {
    "date": 1505439809.19544,
    "log": "type=USER_AVC msg=audit(1483365608.335:311): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='avc:  received policyload notice (seqno=4)  exe=\"/usr/lib/systemd/systemd\" sauid=0 hostname=? addr=? terminal=?'\u001dUID=\"root\" AUID=\"unset\" SAUID=\"root\""
  }
]
```

are you using the latest Fluent Bit v0.12.1 ?

**Comment by jbhilton11** [comment]
It appears this issue was fixed with the latest v0.12.1. Originally I was working with v0.11 and had to strip that value out before trying to parse as JSON, but I commented that portion out and upgraded to v0.12.1 and don't seem to have an issue with that any more. Thanks for your help!

**Comment by edsiper** [comment]
awesome, please make sure to use 0.12.2 (released last week).
